### PR TITLE
osd: allow eviction on non-legacy watched object

### PR DESCRIPTION
**Pull request by wonzhq**
For v2 watch, the watchers are disconnected while doing eviction.
Clients can reconnect to the object in the base. This is to support
removing a cache tier on active RBD pools where the RBD image header
is watched by active clients.

Fixes: http://tracker.ceph.com/issues/14865
Signed-off-by: Zhiqiang Wang zhiqiang@xsky.com

### CODE REVIEW

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -4593,10 +4593,23 @@ int ReplicatedPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
...
 	  break;
 	}
-	if (!oi.watchers.empty()) {
+
+	// return EBUSY if we have legacy watcher. For v2 watch, the watchers
+	// are disconnected in _delete_oid, clients will reconnect to the object
+	// in the base after the eviction.
+	bool have_legacy_watcher = false;
+        for (map<pair<uint64_t, entity_name_t>, WatchRef>::iterator it =
+	     ctx->obc->watchers.begin(); it != ctx->obc->watchers.end(); ++it) {
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: for (auto w : ctx->obc->watchers)

**Comment by liewegas** [comment]
I wonder if we can do better.  For example, the disconnect we queue in _delete_oid passes true for send_disconnect.  If we don't send a disconnect, it could try to reconnect to the same object in the base tier.... except that we force promotion for watch ops.

Hrm.. in order for somethign like that to work, I think the base tier would need to disconnect watchers when the object is promoted, so that we can be sure all watches are on the same tier?

**Comment by liewegas** [comment]
In the meantime, this is probably better than nothing... but...

**Comment by wonzhq** [comment]
@liewegas I agree with that the base tier needs to disconnect watchers when the object is promoted. But when evicting an object, isn't it better to send disconnects to all the watchers, so that they can reconnect to the base object ASAP (we are removing the cache tier, and the cache mode has been changed to forward in this case)?

**Comment by liewegas** [comment]
Hmm, you're right that we need to send something, since there isn't a PG 
change to make the client reconnect.  But a disconnect like this is an 
error that surfaces up to the librados user, and I think what we actually 
want is for the objecter to do it's internal reconnect attempt (like it 
does if the PG moves), and surface an error only if that fails.

I guess this is fine in the interim, since well-written clients will 
try to reconnect and revalidate all of their state.  But it's more violent 
than it should be.

We probably need a new event type CEPH_WATCH_EVENT_MUST_RECONNECT or 
something...

**Comment by wonzhq** [comment]
Got you. Yes it's better to come up with a new event type.

**Comment by liewegas** [comment]
@athanatos what do you think we should do here?

1) merge this as is.  it'll let you cache-flush an rbd header, and librbd should reconnect.
2) extend the watch/notify protocol so that there is a graceful way to ask the client to reconnect.

I think 2 will get messy.

**Comment by liewegas** [comment]
@wonzhq I think we need to add a new test/librados/tier.cc test that creates an object in teh cache tier, sets a watch, and does a cache-flush and evict, and verifies evict returns 0 and the watch  gets disconnected.

**Comment by wonzhq** [comment]
@liewegas, sorry I wasn't able to work on this in the last couple of days for some reason. So we are going ahead with option 1? Personally I prefer option 2. Anyway, I'll add the test case ASAP.

**Comment by wonzhq** [comment]
@liewegas test added, pls take another look, thx!

**Comment by athanatos** [comment]
We need a test that verifies rbd's behavior as well.

**Comment by athanatos** [comment]
Option 1 is I think adequate for now.  If we want to do option 2, first ste

**Comment by athanatos** [comment]
Oops, didn't mean to close.

**Comment by athanatos** [comment]
@wonzhq We still need an rbd test.

**Comment by dillaman** [comment]
@wonzhq ping

**Comment by wonzhq** [comment]
Sorry, I'm not sure how to add the rbd test. Pls advise..

**Comment by athanatos** [comment]
@dillaman What would @wonzhq need to do to get rbd to exercise the code?

**Comment by dillaman** [comment]
@wonzhq If you could write a bash script in "qa/workunits/rbd" that creates an image, runs "rbd watch" against the image in the background, ensure the image header is promoted to the cache pool, demote it, and run "rbd resize" against the image and verify that your background "rbd watch" sees the notification.

**Comment by liewegas** [comment]
@wonzhq ping

**Comment by liewegas** [comment]
please reopen when you have time to address this!

**Pull request closed without merging**